### PR TITLE
test runner

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
This PR contains the following:
- `test` attribute for functions and appropriate checks
- test function compilation
- `test-runner`  crate with revm backend
- `tests` crate which uses the test runner (renamed old tests crate to `tests-legacy`)
- CLI `test` subcommand

Example CLI usage:

Screenshot from 2023-02-28 16-14-28

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe [hunk 1]
```diff
@@ -13,3 +14,17 @@ pub contract BingContract {
         return x + y
     }
 }
+
+#test
+fn get_42_backend_is_41() {
+    assert get_42_backend() == 41
+}
+
+// #test(ctx: TestContext)
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe:23]: delete

File: crates/tests/src/ingots.rs [hunk 2]
```diff
@@ -50,16 +50,16 @@ fn test_basic_ingot() {
...
-            &mut executor,
-            "get_my_dyng",
-            &[],
-            Some(&tuple_token(&[
-                address_token("8"),
-                uint_token(42),
-                int_token(-1),
-            ])),
-        );
+        // harness.test_function(
```
Comment (g-r-a-n-t) [inline crates/tests/src/ingots.rs:53]: readd

File: crates/tests/src/internal.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use insta::assert_snapshot;
+use test_files::fixture;
+
+macro_rules! test_report {
```
Comment (g-r-a-n-t) [inline crates/tests/src/internal.rs:4]: complete these tests and try refactoring some existing tests

File: crates/codegen/src/yul/isel/function.rs [hunk 4]
```diff
@@ -34,6 +34,7 @@ use crate::{
     },
 };
 
+// here
```
Comment (g-r-a-n-t) [inline crates/codegen/src/yul/isel/function.rs:37]: delete

File: crates/driver/Cargo.toml [hunk 5]
```diff
@@ -23,3 +23,9 @@ indexmap = "1.6.2"
 vfs = "0.5.1"
 smol_str = "0.1.21"
 toml = "0.5.8"
+foundry-evm = { git="https://github.com/foundry-rs/foundry" }
+foundry-config = { git="https://github.com/foundry-rs/foundry" }
```
Comment (g-r-a-n-t) [inline crates/driver/Cargo.toml:27]: delete

File: crates/driver/src/lib.rs [hunk 6]
```diff
@@ -1,21 +1,34 @@
-#![allow(unused_imports, dead_code)]
+#!allow(unused_imports, dead_code)]
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:1]: ?

File: crates/driver/src/lib.rs [hunk 7]
```diff
@@ -32,6 +45,110 @@ pub struct CompiledContract {
...
+            sink.insert_failure(&self.name, &format!("{:?}", result.exit_reason));
+        } else {
+            sink.inc_success_count();
+        };
+        !result.reverted
+    }
+}
+
+#[derive(Debug, Default)]
+// #[cfg(feature = "solc-backend")]
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:88]: remove solc-backend feature constraint

File: crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe (hunk 1 above)
Comment (g-r-a-n-t) [reply crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe:23]: just remove this stuff

File: crates/test-files/fixtures/ingots/basic_ingot/src/main.fe [hunk 8]
```diff
@@ -37,3 +37,13 @@ contract Foo {
         return bing_contract.add(40, 50)
     }
 }
+
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/ingots/basic_ingot/src/main.fe:40]: also remove

File: crates/test-files/fixtures/internal/const_local.fe [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    const C4: i64 = (C1 * C2) + C3 // 12
+    const EXP: u8 = 2
+    const C5: i64 = (C2 ** EXP) // 4
+    const C6: i64 = (C5 << 1) // 8
+    const C7: bool = (C5 < C6) and (C4 > C5) // true
+    const C8: bool = false
+    const C9: bool = C8 < C7 // true
+    const C10: u256 = 42 if C9 else 0 // 42
+    // todo: why does this fail?
+    // let _arr: Array<bool, { C10 }> = [true; { C10 }]
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/internal/const_local.fe:14]: readd

File: crates/test-files/fixtures/internal/revert.fe [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#test
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/internal/revert.fe:1]: delete

File: crates/test-files/fixtures/internal/while_loop.fe [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    let mut val: u256 = 0
+
+    while val < 2 {
+        val = val + 1
+        if val == 2 {
+            val = 3
+        }
+    }
+
+    // let val2: u256 = 3
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/internal/while_loop.fe:11]: delete

File: crates/tests/src/internal.rs [hunk 12]
```diff
@@ -0,0 +1,46 @@
+use fe_driver::TestSink;
+use test_files::{fixture, fixture_dir_files};
+
+macro_rules! test_report {
+    // ($name:ident) => {
```
Comment (g-r-a-n-t) [inline crates/tests/src/internal.rs:5]: readd

File: todo [hunk 13]
```diff
@@ -0,0 +1,5 @@
+- add module name to test print
```
Comment (g-r-a-n-t) [inline todo:1]: complete and remove

File: crates/driver/Cargo.toml [hunk 14]
```diff
@@ -19,7 +19,11 @@ fe-mir = {path = "../mir", version = "^0.21.0-alpha"}
...
 fe-parser = {path = "../parser", version = "^0.21.0-alpha"}
 fe-yulc = {path = "../yulc", version = "^0.21.0-alpha", features = ["solc-backend"], optional = true}
+fe-test-runner = {path = "../test-runner", version = "^0.21.0-alpha"}
 indexmap = "1.6.2"
 vfs = "0.5.1"
 smol_str = "0.1.21"
 toml = "0.5.8"
+primitive-types = "0.12"
+colored = "2.0.0"
+hex="0.4"
```
Comment (g-r-a-n-t) [inline crates/driver/Cargo.toml:29]: check if these are still needed

File: crates/fe/Cargo.toml [hunk 15]
```diff
@@ -19,7 +19,8 @@ fs_extra = "1.2.0"
 walkdir = "2"
 indexmap = "1.6.2"
 include_dir = "0.7.2"
-
+colored = "2.0.0"
```
Comment (g-r-a-n-t) [inline crates/fe/Cargo.toml:22]: might not be needed

File: crates/tests-legacy/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,42 @@
+[package]
```
Comment (g-r-a-n-t) [inline crates/tests-legacy/Cargo.toml:1]: try to fix git here

File: crates/driver/src/lib.rs [hunk 17]
```diff
@@ -127,13 +193,33 @@ pub fn dump_mir_single_file(db: &mut Db, path: &str, src: &str) -> Result<String
 }
 
 #[cfg(feature = "solc-backend")]
-fn compile_module_id(
+fn compile_test(db: &mut Db, test: FunctionId, optimize: bool) -> CompiledTest {
+    let yul_test = fe_codegen::yul::isel::lower_test(db, test)
+        .to_string()
+        .replace('"', "\\\"");
+    let bytecode = compile_to_evm("test", &yul_test, optimize);
+    // panic!("{}\n{}", yul_test, bytecode);
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:201]: delete

**Comment by g-r-a-n-t** [comment]
I'm using revm now @sbillig. There's a nested dependency on `getrand` that fails to build for wasm. I'm not sure how to easily fix this, so I'm just disabling revm for wasm. let me know if you have any ideas

File: crates/analyzer/tests/errors.rs [hunk 18]
```diff
@@ -355,5 +355,5 @@ test_file! { invalid_struct_pub_qualifier }
 test_file! { mut_mistakes }
 test_file! { invalid_comparisons }
 
-test_file! { test_call }
-test_file! { test_params }
+// test_file! { test_call }
+// test_file! { test_params }
```
Comment (sbillig) [inline crates/analyzer/tests/errors.rs:359]: @g-r-a-n-t These tests aren't producing snapshots, which breaks the test on wasm. I assume there's no error output? (The `test_file` macro should probably assert that there's error output)

File: .github/workflows/main.yml [hunk 19]
```diff
@@ -98,13 +98,11 @@ jobs:
         run: curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
       - name: Cache Dependencies
         uses: Swatinem/rust-cache@v2
-      - name: Build WASM tests
-        run: wasm-pack test --node crates/fe -- --workspace --no-run
       - name: Run WASM tests
         # wasm-pack needs a Cargo.toml with a 'package' field.
         # (see https://github.com/rustwasm/wasm-pack/issues/642)
         # This will still run all tests in the workspace.
-        run: wasm-pack test --node crates/fe -- --workspace
```
Comment (sbillig) [inline .github/workflows/main.yml:107]: I'm pretty sure this used to run the tests, but now doesn't, so we haven't actually been running wasm tests for a while (and they've been broken). I guess it was a wasm-pack change? :shrug:

**Comment by sbillig** [comment]
@g-r-a-n-t I pushed a commit to let the test-runner crate build on wasm, but of course we can't run any of the tests because we can't build the solc backend on wasm :facepalm:. Anyway, I still think this was a worthwhile exercise.

File: crates/analyzer/tests/errors.rs (hunk 18 above)
Comment (g-r-a-n-t) [reply crates/analyzer/tests/errors.rs]: yeah, looks like insta just does this

```
#[test]
fn test_my_thing() {
    assert_snapshot!("test123")
}
```

```
Snapshot file: crates/analyzer/tests/snapshots/errors__my_thing.snap
Snapshot: my_thing
Source: crates/analyzer/tests/errors.rs:101
```

**Review by sbillig — APPROVED** [review]
Looks good

**Pull request merged**